### PR TITLE
flto gcc flag only for data.table

### PR DESCRIPTION
**Pull request by jangorecki**
follow up of #3878

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -139,8 +138,8 @@ test-rel-cran-lin: # currently released R on Linux
     _R_CHECK_CRAN_INCOMING_REMOTE_: "FALSE"   # Other than no URL checking (takes many minutes) or 'Days since last update 0' NOTEs needed, #3284
   before_script:
     - mkdir -p ~/.R
-    - echo 'CFLAGS=-g0'> ~/.R/Makevars   # No -g because -g increases datatable.so size from 0.5MB to 1.5MB and breaches 'installed package size <= 5MB' note
-    - echo 'CXXFLAGS=-g0' >> ~/.R/Makevars
+    - echo 'CFLAGS=-g0 -O2 -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2'> ~/.R/Makevars # -g0 because -g increases datatable.so size from 0.5MB to 1.5MB and breaches 'installed package size <= 5MB' note
```
Comment (jangorecki) [inline .gitlab-ci.yml]: these are the defaults flags, they were disabled when writing just `-g0` to Makevars

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -100,15 +100,14 @@ test-rel-lin: #  most comprehensive tests, force all suggests, also integration
...
-    - echo 'CFLAGS=-g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2' > ~/.R/Makevars
-    - echo 'CXXFLAGS=-g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2' >> ~/.R/Makevars
   script:
     - Rscript -e 'source("ci.R"); install.packages(dcf.dependencies(c("DESCRIPTION","inst/tests/tests-DESCRIPTION"), which="all"))'
     - *copy-src
     - rm -r bus
     - *move-src
+    - mkdir -p ~/.R
+    - echo 'CFLAGS=-g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2' > ~/.R/Makevars
+    - echo 'CXXFLAGS=-g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2' >> ~/.R/Makevars
```
Comment (jangorecki) [inline .gitlab-ci.yml]: this is the fix for current failure due to `-flto` applied to jsonlite (and others) that fail when compiling with this flag

**Comment by mattdowle** [comment]
I added the 1.12.4 milestone,  but then removed it because this isn't the package.  So it might be wasteful to pollute the tag with non-package tasks.  On the other hand we're using the milestone tag to mark what's going ahead now, or not.  Maybe the rule should be if it's tagged CI then we shouldn't mark the milestone.

**Comment by jangorecki** [comment]
I think it is fine to tag milestone for stuff that is in .Rbuildignore

**Comment by jangorecki** [comment]
Even when having such a rule, there is no way to align to that well. It is easier to have a rule the other way around.

**Comment by jangorecki** [comment]
PR fixed previous `flto` issue: https://gitlab.com/Rdatatable/data.table/-/jobs/296558457

----

failure state is now due to grep, for some reason, ci yaml line
```
! grep "warning:" data.table.Rcheck/00install.out
```
is run in a job without leading `!`
```
grep "warning:" data.table.Rcheck/00install.out
```

----

[00install.log](https://gitlab.com/Rdatatable/data.table/-/jobs/296558457/artifacts/raw/bus/test-rel-lin/data.table.Rcheck/00install.out)
```
* installing *source* package ‘data.table’ ...
** using staged installation
** libs
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c assign.c -o assign.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c between.c -o between.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c bmerge.c -o bmerge.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c chmatch.c -o chmatch.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c cj.c -o cj.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c coalesce.c -o coalesce.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c dogroups.c -o dogroups.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c fastmean.c -o fastmean.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c fcast.c -o fcast.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c fifelse.c -o fifelse.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c fmelt.c -o fmelt.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c forder.c -o forder.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c frank.c -o frank.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c fread.c -o fread.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c freadR.c -o freadR.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c froll.c -o froll.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c frollR.c -o frollR.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c frolladaptive.c -o frolladaptive.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c fsort.c -o fsort.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c fwrite.c -o fwrite.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c fwriteR.c -o fwriteR.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c gsumm.c -o gsumm.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c ijoin.c -o ijoin.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c init.c -o init.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c inrange.c -o inrange.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c nafill.c -o nafill.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c nqrecreateindices.c -o nqrecreateindices.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c openmp-utils.c -o openmp-utils.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c quickselect.c -o quickselect.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c rbindlist.c -o rbindlist.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c reorder.c -o reorder.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c shift.c -o shift.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c subset.c -o subset.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c transpose.c -o transpose.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c types.c -o types.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c uniqlist.c -o uniqlist.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c utils.c -o utils.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c vecseq.c -o vecseq.o
gcc -std=gnu99 -I"/usr/share/R/include" -DNDEBUG    -fopenmp -fpic  -g -O3 -flto -fopenmp -Wall -pedantic -fstack-protector-strong -D_FORTIFY_SOURCE=2 -c wrappers.c -o wrappers.o
gcc -std=gnu99 -shared -L/usr/lib/R/lib -Wl,-Bsymbolic-functions -Wl,-z,relro -o data.table.so assign.o between.o bmerge.o chmatch.o cj.o coalesce.o dogroups.o fastmean.o fcast.o fifelse.o fmelt.o forder.o frank.o fread.o freadR.o froll.o frollR.o frolladaptive.o fsort.o fwrite.o fwriteR.o gsumm.o ijoin.o init.o inrange.o nafill.o nqrecreateindices.o openmp-utils.o quickselect.o rbindlist.o reorder.o shift.o subset.o transpose.o types.o uniqlist.o utils.o vecseq.o wrappers.o -fopenmp -lz -L/usr/lib/R/lib -lR
mv data.table.so datatable.so
if [ "" != "Windows_NT" ] && [ `uname -s` = 'Darwin' ]; then install_name_tool -id datatable.so datatable.so; fi
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/assign.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/between.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/bmerge.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/chmatch.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/cj.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/coalesce.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/dogroups.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/fastmean.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/fcast.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/fifelse.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/fmelt.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/forder.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/frank.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/fread.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/freadR.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/froll.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/frollR.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/frolladaptive.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/fsort.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/fwrite.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/fwriteR.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/gsumm.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/ijoin.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/init.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/inrange.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/nafill.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/nqrecreateindices.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/openmp-utils.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/quickselect.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/rbindlist.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/reorder.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/shift.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/subset.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/transpose.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/types.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/uniqlist.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/utils.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/vecseq.o: plugin needed to handle lto object
/usr/bin/nm: /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00_pkg_src/data.table/src/wrappers.o: plugin needed to handle lto object
installing to /builds/Rdatatable/data.table/bus/test-rel-lin/data.table.Rcheck/00LOCK-data.table/00new/data.table/libs
** R
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
** building package indices
** installing vignettes
** testing if installed package can be loaded from temporary location
** checking absolute paths in shared objects and dynamic libraries
** testing if installed package can be loaded from final location
** testing if installed package keeps a record of temporary installation path
* DONE (data.table)
```

File: .gitlab-ci.yml (hunk 2 above)
Comment (MichaelChirico) [reply .gitlab-ci.yml]: Is this something we should file an issue with for those pkg?

Comment (mattdowle) [reply .gitlab-ci.yml]: They should already know via their CRAN check pages, like data.table shows the red LTO fail currently.

Comment (MichaelChirico) [reply .gitlab-ci.yml]: OK, in fact yes:

Screenshot 2019-09-17 at 2 15 01 PM

(`jsonlite` here)

**Pull request merged**